lib/rjgit.rb

```
@@ -341,21 +334,116 @@ def commit(message, author, parents = nil, ref = nil, force = false)
+      import org.eclipse.jgit.diff.DiffEntry
+
+      ADD    = org.eclipse.jgit.diff.DiffEntry::ChangeType::ADD
+      COPY   = org.eclipse.jgit.diff.DiffEntry::ChangeType::COPY
+      MODIFY = org.eclipse.jgit.diff.DiffEntry::ChangeType::MODIFY
+      DELETE = org.eclipse.jgit.diff.DiffEntry::ChangeType::DELETE
+      RENAME = org.eclipse.jgit.diff.DiffEntry::ChangeType::RENAME
+
+      # Take the result of RJGit::Porcelain.diff with options[:patch] = true and return a patch String
+      def self.diffs_to_patch(diffs)
+        result = ""
+        diffs.each {|diff| result << diff[:patch]}
+        result
+      end
+
+      def initialize(repository, patch, ref = Constants::HEAD)
+        super(repository)
+        @ref   = ref
+        @patch = Patch.new
+        @patch.parse(ByteArrayInputStream.new(patch.to_java_bytes))
+        raise_patch_apply_error unless @patch.getErrors.to_a.empty?
+        @current_tree = Commit.find_head(@jrepo, ref).tree
+      end
+
+      def commit(message, author, parents = nil, force = false)
+        super(message, author, parents, @ref, force)
+      end
+
+      def build_map
+        raise_patch_apply_error if @patch.getFiles.to_a.empty?
+        @patch.getFiles.each do |file_header|
+          case file_header.getChangeType
+          when ADD
+            add(file_header.getNewPath, apply('', file_header))
+          when MODIFY
+            add(file_header.getOldPath, apply(getData(file_header.getOldPath), file_header))
+          when DELETE
+            delete(file_header.getOldPath)
+          when RENAME
+            delete(file_header.getOldPath)
+            add(file_header.getNewPath, getData(file_header.getOldPath))
+          when COPY
+            add(file_header.getNewPath, getData(file_header.getOldPath))
+          end
+        end
+      end
+
+      private
+
+      def raise_patch_apply_error
+        raise Java::OrgEclipseJgitApiErrors::PatchApplyException.new('Patch failed to apply')
```

Does it make sense to raise a JGit Exception here, or should we raise an RJGit exception instead? (Since we are in the business of wrapping JGit.)

---

Good point. `RJGit::Porcelain.apply` would currently throw a JGit Exception, so I thought I would mimic that behavior. Or should we throw an RJGit exception in both places?

---

Throwing an RJGit exception in both places.